src/components/Dropdown.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useCallback, useMemo, useState } from 'react';
+import { Person } from '../types/Person';
+import debounce from 'lodash.debounce';
+import { peopleFromServer } from '../data/people';
+
+type Props = {
+  onSelected: (person: Person | null) => void;
+  delay: number;
+  selectedPerson: Person | null;
+};
+
+export const Dropdown: React.FC<Props> = ({
+  onSelected,
+  delay,
+  selectedPerson,
+}) => {
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
+  const [listOfPeople, setListOfPeople] = useState(peopleFromServer);
+  const [listVisibility, setListVisibility] = useState(false);
+
+  const filteredPeopleList = useMemo(() => {
+    if (!appliedQuery) {
+      return peopleFromServer;
+    }
+
+    return peopleFromServer.filter((person: Person) =>
+      person.name.toLowerCase().includes(appliedQuery.toLowerCase()),
+    );
+  }, [appliedQuery, listOfPeople]);
+
+  const applyQuery = useCallback(debounce(setAppliedQuery, delay), []);
+
+  const handleQueryChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const inputValue = event.target.value;
+
+    setQuery(inputValue);
+    setListVisibility(true);
+
+    if (selectedPerson && inputValue !== selectedPerson.name) {
+      onSelected(null);
+    }
+
+    applyQuery(inputValue);
+    setListOfPeople(filteredPeopleList);
```

Comment:
you don't need a separate state for it
just use variable filteredPeopleList - it will be recalculated every time when related state variable change, so re-render will happen and correct list will be displayed